ports/unix/mphciport.c

```
@@ -0,0 +1,198 @@
+
+STATIC int uart_fd = -1;
+STATIC pthread_t hci_poll_thread_id;
+
+STATIC void *hci_poll_thread(void *arg) {
+    (void)arg;
+
+    // This will return false when the stack is shutdown.
+    while (mp_bluetooth_hci_poll()) {
+        usleep(UART_POLL_INTERVAL_US);
+    }
+
+    return NULL;
+}
+
+STATIC int configure_uart(void) {
+    struct termios toptions;
+
+    // Get existing config.
+    if (tcgetattr(uart_fd, &toptions) < 0) {
+        DEBUG_printf("Couldn't get term attributes");
+        return -1;
+    }
+
+    // Raw mode (disable all processing).
+    cfmakeraw(&toptions);
+
+    // 8N1, no parity.
+    toptions.c_cflag &= ~CSTOPB;
+    toptions.c_cflag |= CS8;
+    toptions.c_cflag &= ~PARENB;
+
+    // Enable receiver, ignore modem control lines
+    toptions.c_cflag |= CREAD | CLOCAL;
+
+    // Blocking, single-byte reads.
+    toptions.c_cc[VMIN] = 1;
+    toptions.c_cc[VTIME] = 0;
+
+    // Enable HW RTS/CTS flow control.
+    toptions.c_iflag &= ~(IXON | IXOFF | IXANY);
+    toptions.c_cflag |= CRTSCTS;
+
+    // 1Mbit (TODO: make this configurable).
+    speed_t brate = B1000000;
+    cfsetospeed(&toptions, brate);
+    cfsetispeed(&toptions, brate);
+
+    // Apply immediately.
+    if (tcsetattr(uart_fd, TCSANOW, &toptions) < 0) {
+        DEBUG_printf("Couldn't set term attributes");
+        return -1;
+    }
+
+    return 0;
+}
+
+// HCI UART bindings.
+int mp_bluetooth_hci_uart_init(uint32_t port, uint32_t baudrate) {
+    (void)port;
+    (void)baudrate;
+
+    DEBUG_printf("mp_bluetooth_hci_uart_init (unix)\n");
+
+    char uart_device_name[256] = "/dev/ttyUSB0";
+
+    char *path = getenv("MICROPYBTUART");
+    if (path != NULL) {
+        strcpy(uart_device_name, path);
+    }
+    DEBUG_printf("Using HCI UART: %s\n", uart_device_name);
+
+    int flags = O_RDWR | O_NOCTTY | O_NONBLOCK;
+    uart_fd = open(uart_device_name, flags);
+    if (uart_fd == -1) {
+        DEBUG_printf("Unable to open port %s", uart_device_name);
```

This would be better as an exception that's then user visible - currently if the port is missing or cannot be opened it's a silent fail here and the rest of the startup just looks broken

---

Done -- and added the plumbing to make this work for btstack and nimble. On btstack it can detect the init failure, on nimble it now leads to an init timeout.